### PR TITLE
[libc++] Fix acceptance of convertible-to-{float,double,long double} in std::isfinite()

### LINKED ISSUE #98816: [libc++] std::isfinite() does not accept convertible-to-float
The following code works with libstdc++, fails with libc++:
```c++
#include <cmath>

struct X {
    operator float();
};

bool test(X value) {
    return std::isfinite(value);
}
```

Conformance viewer: https://godbolt.org/z/nK9sofToP

Long time ago probably the same issue was reported in https://github.com/llvm/llvm-project/issues/23081 and closed as a duplicate of https://github.com/llvm/llvm-project/issues/18592 . However the fix was applied only for `isnan` and `isinf` - see https://github.com/llvm/llvm-project/blob/fe7fc54403e17b17e366adaab75256b2269fc204/libcxx/include/__math/traits.h#L97-L110

Could you apply the same fix for `std::isfinite`?

### PR DESCRIPTION
**Pull request by robincaloudis**
Closes https://github.com/llvm/llvm-project/issues/98816.

### CODE REVIEW

**Comment by AngryLoki** [comment]
Fails in gcc test... Looks like std::isnan `_LIBCPP_PREFERRED_OVERLOAD` also never worked in gcc too and it was not covered with tests. I guess it is possible to implement in in gcc (https://godbolt.org/z/vcrGWzcvj). But I don't know if this level of complication is needed. libstdc++ has no such thing as "preferred overload" and just fails with `call to 'isfinite' is ambiguous` when there are multiple user-defined conversion functions.

**Comment by ldionne** [comment]
Thanks for the patch! Is there a reason why we must apply `_LIBCPP_PREFERRED_OVERLOAD` at all? This might be a naive question, but GCC implements the same thing without any trickery so I wonder why we need `_LIBCPP_PREFERRED_OVERLOAD` in the first place.

File: libcxx/test/std/numerics/c.math/isfinite.pass.cpp
Comment (philnik777) [inline libcxx/test/std/numerics/c.math/isfinite.pass.cpp:82]:
```suggestion
template <class T>
struct ConvertibleTo {
  operator T() const { return {}; }
};
```
And then use `ConvertibleTo<float>` below.

**Review by philnik777 — COMMENTED** [review]
Could you also check the other traits to make sure they're correctly implemented?

**Comment by robincaloudis** [comment]
> Could you also check the other traits to make sure they're correctly implemented?

Sure, I'll check them.

File: libcxx/test/std/numerics/c.math/isfinite.pass.cpp
Comment (ldionne) [inline libcxx/test/std/numerics/c.math/isfinite.pass.cpp:88]:
```suggestion
  // Make sure we can call `std::isfinite` with convertible-to-XXX types
  {
    assert(std::isfinite(ConvertibleFloat(0)));
    assert(std::isfinite(ConvertibleDouble(0)));
    assert(std::isfinite(ConvertibleLongDouble(0)));
  }
```

**Review by ldionne — APPROVED** [review]
LGTM once CI is done running. MacOS is a bit backed up right now due to an outage but it should get there eventually. Thanks!

**Comment by ldionne** [comment]
Sorry, @philnik777 and I got in a race condition. Please make sure to satisfy Nikolas' requests as well. Thanks :-)

**Review by ldionne — APPROVED** [review]
Can be merged once CI is done running.

**Comment by robincaloudis** [comment]
CI is done running. Could you merge this PR? I lack write access. Thanks!

**Comment by AngryLoki** [comment]
That was quick! Thanks everyone!

**Pull request merged**
